gtk4/sample/examples/application8/exampleapp.rb

```diff
@@ -0,0 +1,358 @@
+      set_template(data: template)
+      bind_template_child("stack")
+      bind_template_child("gears")
+      bind_template_child("search")
+      bind_template_child("searchbar")
+      bind_template_child("searchentry")
+      bind_template_child("words")
+      bind_template_child("sidebar")
+      bind_template_callback_full("search_text_changed")
+      bind_template_callback_full("visible_child_changed")
+    end
+  end
+
+  def initialize(application)
+    menu_ui = <<~MENU
+    <?xml version="1.0" encoding="UTF-8"?>
+    <interface>
+      <menu id="menu">
+        <section>
+          <item>
+            <attribute name="label" translatable="yes">_Words</attribute>
+            <attribute name="action">win.show-words</attribute>
+          </item>
+          <item>
+            <attribute name="label" translatable="yes">_Preferences</attribute>
+            <attribute name="action">app.preferences</attribute>
+          </item>
+        </section>
+        <section>
+          <item>
+            <attribute name="label" translatable="yes">_Quit</attribute>
+            <attribute name="action">app.quit</attribute>
+          </item>
+        </section>
+      </menu>
+    </interface>
+    MENU
+    super(application: application)
+    builder = Gtk::Builder.new(string: menu_ui)
+    gears.menu_model = builder["menu"]
+    @settings = Gio::Settings.new("org.gtk.exampleapp")
+    @settings.bind("transition", stack, "transition-type", :default)
+    @settings.bind("show-words", sidebar, "reveal-child", :default)
+    search.bind_property("active", searchbar, "search-mode-enabled", :bidirectional)
+    sidebar.signal_connect "notify::reveal-child" do |_sidebar, _pspec|
+      update_words
+    end
+    action = @settings.create_action("show-words")
+    add_action(action)
+end
+
+  def open(file)
+    basename = file.basename
+    scrolled = Gtk::ScrolledWindow.new
+    scrolled.hexpand = true
+    scrolled.vexpand = true
+    view = Gtk::TextView.new
+    view.editable = false
+    view.cursor_visible = false
+    scrolled.child = view
+    stack.add_titled(scrolled, basename, basename)
+    buffer = view.buffer
+    stream = file.read
+    buffer.text = stream.read
+    tag = buffer.create_tag
+    @settings.bind("font", tag, "font", :default)
+    buffer.apply_tag(tag, buffer.start_iter, buffer.end_iter)
+    search.sensitive = true
+    update_words
+  end
+
+  private
+
+  def search_text_changed(search_entry)
+    text = search_entry.text
+    return if text.empty?
+
+    tab = stack.visible_child
+    view = tab.child
+    buffer = view.buffer
+    range = buffer.start_iter.forward_search(text, :case_insensitive)
+    return unless range
+    buffer.select_range(range[0], range[1])
+    view.scroll_to_iter(range[0], 0.0, false, 0.0, 0.0)
+  end
+
+  def update_words
+    tab = stack.visible_child
+    return unless tab
+    view = tab.child
+    buffer = view.buffer
+    strings = buffer.text.downcase.split(/\W+/).uniq
```

Comment:
`\W+` doesn't work with non-ASCII characters such as Japanese word. How about using the default separator?

```suggestion
    strings = buffer.text.downcase.split.uniq
```

Reply:
The original program picks up "words" which are defined in Pango.
And Pango supports Unicode.
Words don't include ";",  ":" or some other characters.
The default argument of the split method is `$;` and `$;` is usually nil.
Then split uses space characters as delimiters, which don't include ";" and ":".
Therefore, the behavior is different.
For example,

Sample text:
abc def; ghi.

Original program:
=> ["abc", "def", "ghi"]

Split method with the default argument:
=> ["abc", "def;", "ghi."]

I think maybe the following code works in Unicode texts.

```ruby
strings = buffer.text.downcase.split(/[^[:word:]]+/).uniq
```

How about it?

Reply:
I found that `example.app` didn't support Japanese text.
Maybe we need to modify:

`stream.read` => `stream.read.force_encoding(Encoding::UTF_8)`

Reply:
`strings = buffer.text.downcase.scan(/[[:word:]]+/).uniq` is better. We can remove `strings.delete("")` with it.

```ruby
file.read do |stream|
  buffer.text = stream.read.force_encoding(Encoding::UTF_8)
end
```

is better.